liquibase-core/src/main/java/liquibase/sqlgenerator/core/CreateProcedureGenerator.java

```diff
@@ -167,4 +189,109 @@ public static String addSchemaToText(String procedureText, String schemaName, St
         }
         return procedureText;
     }
+
+    /**
+     * Split <code>SET ANSI_NULLS/QUOTED_IDENTIFIER ON/OFF</code> statements out of SQL statement
+     * into before statements procedure/function/trigger/view body and adds SET ANSI_NULLS ON and
+     * SET QUOTED_IDENTIFIER ON into before list to isolate settings change between scripts
+     * by keeping default ANSI_NULLS and QUOTED_IDENTIFIER settings behaviour.
+     *
+     * @param body - SQL to split if SET statements are present
+     * @param endDelimiter - end line delimiter
+     * @return - before statements if any are present and procedure/function/trigger/view body
+     */
+    public static MssqlSplitStatements splitSetStatementsOutForMssql(String body, String endDelimiter,
+                                                                     List<String> bodyStartStatements) {
+        MssqlSplitStatements mssqlSplitStatements = new MssqlSplitStatements();
+
+        StringClauses sqlClauses = SqlParser.parse(body, true, true);
+        StringClauses.ClauseIterator clauseIterator = sqlClauses.getClauseIterator();
+        Object next = "";
+        ArrayList<String> beforeStatements = new ArrayList<>();
+
+        beforeStatements.add("SET ANSI_NULLS ON");
+        beforeStatements.add("SET QUOTED_IDENTIFIER ON");
+
+        while (next != null && clauseIterator.hasNext()) {
+            next = clauseIterator.nextNonWhitespace();
+
+            next = splitOutIfSetStatement(next, clauseIterator, endDelimiter, beforeStatements);
+
+            if (next != null && BEFORE_BODY_START_STATEMENTS.contains(next.toString().toUpperCase())) {
+                next = clauseIterator.nextNonWhitespace();
+                if (next != null && bodyStartStatements.contains(next.toString().toUpperCase())) {
+                    break;
+                }
+            }
+        }
+
+        mssqlSplitStatements.setSetStatementsBefore(beforeStatements);
+        mssqlSplitStatements.setBody(sqlClauses.toString().trim());
+
+        return mssqlSplitStatements;
+    }
+
+    private static Object splitOutIfSetStatement(Object next, StringClauses.ClauseIterator clauseIterator,
+                                                 String endDelimiter, ArrayList<String> setStatements) {
+        while (next != null && next.toString().equalsIgnoreCase("SET")) {
```

Comment:
```suggestion
        while ("SET".equalsIgnoreCase(next)) {
```

Reply:
"SET".equalsIgnoreCase(next) <- It does not compile, because next is an object
"SET".equalsIgnoreCase((String) next) <- it will compile but we don't have guarantee that next is String, so we might get an error
"SET".equalsIgnoreCase(next.toString()) <- will work, but if it is null, then it will throw NPE, so current check looks optimal